src/notebooks/controllers/controllerPreferredService.ts

```diff
@@ -0,0 +1,318 @@
+                }
+
+                // Send telemetry on looking for preferred don't await for sending it
+                this.sendPreferredKernelTelemetry(
+                    document.uri,
+                    notebookMetadata,
+                    preferredConnection,
+                    preferredInterpreter
+                ).ignoreErrors();
+
+                // If we found a preferred kernel, set the association on the NotebookController
+                if (preferredSearchToken.token.isCancellationRequested && !preferredConnection) {
+                    traceInfo('Find preferred kernel cancelled');
+                    return {};
+                }
+                if (!preferredConnection) {
+                    traceInfoIfCI(
+                        `PreferredConnection not found for NotebookDocument: ${getDisplayPath(document.uri)}`
+                    );
+                    return {};
+                }
+
+                traceInfo(
+                    `PreferredConnection: ${preferredConnection.id} found for NotebookDocument: ${getDisplayPath(
+                        document.uri
+                    )}`
+                );
+
+                const targetController = this.registration.values.find(
+                    (value) => preferredConnection?.id === value.connection.id
+                );
+                // If the controller doesn't exist, then it means we're still loading them.
+                // However we can create this one as we have all of the necessary info.
+                if (!targetController) {
+                    traceVerbose(`Early registration of controller for Kernel connection ${preferredConnection.id}`);
+                    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+                    this.registration.add(preferredConnection, [JupyterNotebookView]);
+                }
+            } else if (document.notebookType === InteractiveWindowView) {
+                // Wait for our controllers to be loaded before we try to set a preferred on
+                // can happen if a document is opened quick and we have not yet loaded our controllers
+                await loadControllersPromise;
+
+                // For interactive set the preferred controller as the interpreter or default
+                const defaultInteractiveController = await this.defaultService.computeDefaultController(
+                    document.uri,
+                    'interactive'
+                );
+                preferredConnection = defaultInteractiveController?.connection;
+            }
+
+            // See if the preferred connection is in our registered controllers, add the sufix for the interactive scenario
+            let targetController;
+            if (preferredConnection) {
+                // eslint-disable-next-line @typescript-eslint/no-explicit-any
+                targetController = this.registration.get(preferredConnection, document.notebookType as any);
```

Comment:
Same here, don't we need to controllers to wait until they have loaded?

Reply:
If the two code paths above both wait, then no.

Reply:
Yes, but if they don't, e.g. in the case of a remote kernels (first time opening a new notebook) or the like.
E.g. assume a user doesn't have any local kernels, then looking at this code path we don't wait for remote kernels to finish loading. 

Or Do we? Based on the code it doesn't look like we do.